rust/windows-client/src-tauri/src/client/crash_handling.rs

```diff
@@ -0,0 +1,172 @@
+//! A module for handling crashes and writing minidump files
+//!
+//! Mostly copied from <https://github.com/EmbarkStudios/crash-handling/blob/main/minidumper/examples/diskwrite.rs>
+//!
+//! TODO: Capture crash dumps on panic.
+
+use anyhow::{anyhow, bail, Context};
+use known_folders::{get_known_folder_path, KnownFolder};
+use std::{fs::File, io::Write, path::PathBuf};
+
+const SOCKET_NAME: &str = "dev.firezone.client.crash_handler";
+
+use minidumper::{Client, Server};
+
+/// Attaches a crash handler to the client process
+///
+/// Returns a CrashHandler that must be kept alive until the program exits.
+/// Dropping the handler will detach it.
+#[cfg(debug_assertions)]
+pub(crate) fn attach_handler() -> anyhow::Result<crash_handler::CrashHandler> {
+    attach_handler_inner()
+}
+
+#[cfg(not(debug_assertions))]
+pub(crate) fn attach_handler() -> anyhow::Result<crash_handler::CrashHandler> {
+    // Prevent cargo from complaining
+    if false {
+        return attach_handler_inner();
+    }
+    bail!("crash handling is disabled in release builds for now");
+}
+
+fn attach_handler_inner() -> anyhow::Result<crash_handler::CrashHandler> {
+    // Attempt to connect to the server
+    let (client, _server) = connect()?;
+
+    // Not sure what this does, but the sample code has it.
+    client.send_message(1, "mistakes will be made")?;
+
+    #[allow(unsafe_code)]
+    let handler = crash_handler::CrashHandler::attach(unsafe {
+        crash_handler::make_crash_event(move |crash_context: &crash_handler::CrashContext| {
+            // Before we request the crash, send a message to the server
+            client.send_message(2, "mistakes were made").unwrap();
+
+            // Send a ping to the server, this ensures that all messages that have been sent
+            // are "flushed" before the crash event is sent. This is only really useful
+            // on macos where messages and crash events are sent via different, unsynchronized,
+            // methods which can result in the crash event closing the server before
+            // the non-crash messages are received/processed
+            client.ping().unwrap();
+
+            crash_handler::CrashEventResult::Handled(client.request_dump(crash_context).is_ok())
+        })
+    })
+    .context("failed to attach signal handler")?;
+
+    // On linux we can explicitly allow only the server process to inspect the
+    // process we are monitoring (this one) for crashes
+    #[cfg(any(target_os = "linux", target_os = "android"))]
+    {
+        handler.set_ptracer(Some(_server.id()));
+    }
+
+    Ok(handler)
+}
+
+fn connect() -> anyhow::Result<(Client, std::process::Child)> {
+    let exe = std::env::current_exe().context("unable to find our own exe path")?;
+    let mut server = None;
+
+    for _ in 0..10 {
+        // Create the crash client first so we can error out if another instance of
+        // the Firezone client is already using this socket for crash handling.
```

Comment:
could we set a random socket name so we never overlap?

Reply:
We could, but the client is already supposed to be single-instance because it always has the same name for the wintun adapter, same file path for the device ID, etc.

Reply:
but if we do can we remove this loop?

Reply:
I added a comment saying that basically I don't know why it's there so I don't know if it's okay to remove it